### PR TITLE
[Order Creation] Search Products by SKU

### PR DESCRIPTION
**Pull request by toupper**
Closes: #

## Description

With this PR we add support to search products by SKU when creating an order, something that has been requested by our users a few times.
There have been a few challenges when adapting the SKU search to this case. First and foremost, the products endpoint returns Product Variations as products when returning its data. Because of this, we have to convert it to `ProductVariation` if required, so we can handle them as we were doing before in the `ProductSelectorViewModel`. Furthermore, we show the product variations together with products in the same list as we do in the Products tab, unlike all the other cases in the Product Selector that have a separate screen for all the product variations. This brings more easiness to the user when selecting by SKU, as they can search also variations and select them right away, without having to open a second screen.

Apart from that, we change the logic when the item is deleted from the order detail screen. Before it relied on the product selector to remove it and call back. This raised a bug when the product selector didn't have a reference to the product or variation (e.g. the product page wasn't loaded yet so it wasn't in the database). Now we delete it automatically from the order detail screen. (See changes in the `EditableOrderViewModel`)

Moreover, I took this opportunity to refactor `ProductSelectorViewModel` to collapse all the ID array properties we had into one, encapsulating product and variations alike. Apart from that, I tweaked the sync logic to show the loading screen when we trigger a new search, making it more visible to the users.

## Testing instructions

As the logic for Product Selection changed a bit, please perform a regression test for all the Product Selection cases including editing the order from the order details screen -deleting and changing the quantity of items-. Please also test all the following cases:

- Search SKU for a product and add it. See that the confirm button is updated. Go back to the default state -with no search- and make sure that the product is selected. 
- Search SKU for a variation and add it. See that the confirm button is updated. Go back to the default state -with no search- and make sure that the variation is selected. 
- Deselect products and variations while in SKU search and default mode.

See also that the search is tracked with the right filter:

`🔵 Tracked order_creation_product_selector_search_triggered, properties: [AnyHashable("blog_id"): 205293073, AnyHashable("is_wpcom_store"): true, AnyHashable("search_filter"): "all"]`

and

`🔵 Tracked order_creation_product_selector_search_triggered, properties: [AnyHashable("blog_id"): 205293073, AnyHashable("is_wpcom_store"): true, AnyHashable("search_filter"): "sku"]`

## Screenshots

https://github.com/woocommerce/woocommerce-ios/assets/1864060/0672c32e-7a15-4873-898f-8c3471d1bced

---
- [X] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductSelectorView.swift [hunk 1]
```diff
@@ -1,4 +1,5 @@
 import SwiftUI
+import Yosemite
```
Comment (iamgabrielma) [inline WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductSelectorView.swift:2]: As we're only importing Yosemite for accessing `ProductSearchFilter.allCases`,  should we create an abstraction so we can keep this view decoupled from business logic? Something like:
```
protocol ProductSearchFilterOptions: CaseIterable {
    var title: String { get }
    var analyticsValue: String { get }
}
```

File: WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/SearchHeader.swift [hunk 2]
```diff
@@ -56,7 +60,7 @@ struct SearchHeader: View {
                 .accessibilityHidden(true)
 
             // TextField
-            TextField(placeholder, text: $text)
+            TextField(placeholder, text: $text, onEditingChanged: onEditingChanged ?? { _ in })
```
Comment (iamgabrielma) [inline WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/SearchHeader.swift:63]: TIL about `onEditingChanged` parameter 💯

File: Yosemite/Yosemite/Tools/Products/Product+ProductVariation.swift [hunk 3]
```diff
@@ -0,0 +1,54 @@
+import Foundation
+
+public extension Product {
+    /// In some endpoints, such as the products request with the sku parameters we get Product Variations as products.
+    /// Here we convert it to our ProductVariation entity.
+    ///
+    func toProductVariation() -> ProductVariation? {
```
Comment (iamgabrielma) [inline Yosemite/Yosemite/Tools/Products/Product+ProductVariation.swift:7]: This is a great helper method! 🚀

File: WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductSelectorViewModel.swift [hunk 4]
```diff
@@ -3,6 +3,7 @@ import protocol Storage.StorageManagerType
 import Combine
 import Foundation
 import WooFoundation
+import Networking
```
Comment (iamgabrielma) [inline WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductSelectorViewModel.swift]: Perhaps this import was added automatically by xcode? :D 
```suggestion
```

**Review by iamgabrielma — APPROVED** [review]
Works as expected! 🚢

File: WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductSelectorViewModel.swift (hunk 4 above)
Comment (toupper) [reply WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductSelectorViewModel.swift]: Good catch! Done in 1a4185eff399462aa4aabb4fa92191984ea6a02a

File: WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductSelectorView.swift (hunk 1 above)
Comment (toupper) [reply WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductSelectorView.swift:2]: That's a great question and something that I also considered.
I ended up importing Yosemite as the trade-off is beneficial: if we create the abstraction as you suggested here we will have to convert it back to the Yosemite entity later on, adding a few lines of code that we can spare with this import. As referencing this Yosemite entity from the view doesn't have any clear disadvantage other than what you described, I preferred to leave it so.

**Pull request merged**